---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# SAT



<!-- badges: start -->
<!-- badges: end -->

The goal of SAT is to detect the shared associated analysis. You can download the development version from [GitHub](https://github.com/) with:
                  
```{r}
# install.packages("devtools")
devtools::install_github("JianqiaoWang/SAT")
```


## Prepare LD matrix with the genotype data. 

Option 1: Prepare data for package bigsnpr and bigstatsr

```{r}
library(bigsnpr)
library(bigstatsr)
rds = bigsnpr::snp_readBed2(bedfile = "../../Cric_HF_RealData/rawdata/cric.filtered.maf0.01.bed",
                     backingfile = "cric_geno",ind.col= ind.column)
geno <- snp_attach(rds)
G <- geno$genotypes
CHR <- geno$map$chromosome
infos <- snp_fastImputeSimple(G, method = "mean2")
# To make this permanent, you need to save (modify) the file on disk
geno$genotypes <- infos
geno <- snp_save(geno)
#G <- geno$genotypes
#ind.keep <- snp_clumping(G, infos.chr = geno$map$chromosome,
#                        infos.pos = geno$map$physical.pos,
#                        thr.r2 = 0.01)
#saveRDS(ind.keep, file ="ind_keep_clump.rds")
#data.table::fwrite( as.data.frame(geno$map$marker.ID), file = "snplist.txt")

```



## Example

This is a basic example which shows you how to solve a common problem: Independent sequence

```{r example}
library(SAT)
## basic example code
p <- 1*10^5; ## total number of pairs
d1 = 100; d2 = 100; d3 = 1;
X <- c(rep(0,p-(d1 + d2 + d3)),rep(1,d1),rep(2,d2),rep(3,d3));
mu.1 = rep(0, p); mu.2 = rep(0, p);
mu.1[X==1|X==3] = 4; mu.2[X==2|X==3] = 5;
Z1 <- mu.1 + rnorm(p,0,1);
Z2 <- mu.2 + rnorm(p,0,1);
T1 = abs(Z1);T2 = abs(Z2);
P1 = 2* (1 - pnorm( T1)); P2 = 2* (1 - pnorm(T2));
MinMaxP.simu(P1, P2)
```



```{r}

p <- 1*10^5; ## total number of pairs
d1 = 1000; d2 = 1000; d3 = 100;
X <- c(rep(0,p-(d1 + d2 + d3)),rep(1,d1),rep(2,d2),rep(3,d3));
mu.1 = rep(0, p); mu.2 = rep(0, p);
mu.1[X==1|X==3] = 4; mu.2[X==2|X==3] = 5;
Z1 <- mu.1 + rnorm(p,0,1);
Z2 <- mu.2 + rnorm(p,0,1);
T1 = abs(Z1);T2 = abs(Z2);
P1 = 2* (1 - pnorm( T1)); P2 = 2* (1 - pnorm(T2));
P.min = pmin(P1, P2)
P.max = pmax(P1, P2)
M = min(P.max)
MinMaxP.detect(P.min = P.min, P.max = P.max, M)
MinMaxP.discov(P.max, P.min, method = c("FDR"), alpha = 0.05)

MinMaxP.discov(P.max, P.min, method = c("FWER"), alpha = 0.05)

P.tilde =sapply(P.max, function(x) {Cmax_null(x, P.min ) })

P.fdr = p.adjust(P.tilde, "BH")

which(P.fdr < 0.05)

```



For the typical genetic applications, we consider the LD between the sequences. It requires the additional information of the external LD matrix.  

```{r, eval = F}
pvalue =  SAT::MinMaxP.geno(sumstat.comb$P.gwas, sumstat.comb$P.gtex, ref.geno = geno, ref.bed = "../geno/GTEx_WGS_838Indiv_Freeze_phased_MAF_0.01_chr1_22_EU",
                            output.dir = output.dir, var.name = sumstat.comb$variant_id, block.thresh = 0.99)

```
